docs/DeveloperGuide.adoc

```diff
@@ -111,156 +111,355 @@ image::DeleteSequenceDiagram.png[]
+* Predicates that implements interface `Predicate<Problem>`
++
+These are classes that describes whether an instance of `Problem` is considered a match under a certain field with provided keyword(s).
++
+** `NameContainsKeywordsPredicate`
+*** It ignores case.
+*** It returns true as long as one of the keywords appear in the name as a word. (“As a word” means the matching is done word by word. For instance, `hello` doesn’t match `helloworld`.)
+** `AuthorMatchesKeywordPredicate`
+*** It is case sensitive and matches the entire author string (i.e. requires an exact match).
+** `DescriptionContainsKeywordsPredicate`
+*** It ignores case.
+*** It returns true only when all of the keywords appear in the description as a word.
+** `SourceMatchesKeywordPredicate`
+*** It requires an exact match.
+** `DifficultyIsInRangePredicate`
+*** It matches problems with LOWER_BOUND \<= difficulty \<= UPPER_BOUND
+** `TagIncludesKeywordsPredicate`
+*** Each keyword will be considered as a tag, and two tags are considered equal only when their names are exactly the same.
+*** It returns true when the provided tags are a subset of the tags of the provided problem.
+* `FindCommandParser`
+It validates and parses user input to an instance of `FindCommand`.
 
 [NOTE]
-If the `currentStatePointer` is at index `addressBookStateList.size() - 1`, pointing to the latest address book state, then there are no undone address book states to restore. The `redo` command uses `Model#canRedoAddressBook()` to check if this is the case. If so, it will return an error to the user rather than attempting to perform the redo.
+If the user provides difficulty range as one of the search restrictions, `FindCommandParser` expects the format `LOWER_BOUND \<= difficulty \<= UPPER_BOUND` while `LOWER_BOUND` and `UPPER_BOUND` are valid strings for doubles (i.e. parsable by `Double.parseDouble(...)`).
+
+* `FindCommand`
++
+It creates and stores the `predicate` from an instance of `FindProblemDescriptor`. `predicate` is used to perform the filtering of the displayed problem list when the command is executed.
++
+`predicate` returns true only when the provided problem fulfills all restrictions described by the provided instance of `FindProblemDescriptor`.
++
+Additionally, it implements the following operation(s):
++
+** `FindCommand#execute(...)` - This method overrides `Command#execute(...)`. It filters problems in `filteredProblemList` in `model` with `predicate`.
+** `FindCommand#equals(...)` - Two instances of `FindCommand` are equal if and only if their `predicate` are equal.
+
+image::find/FindCommandClassDiagram.png[]
 
-Step 5. The user then decides to execute the command `list`. Commands that do not modify the address book, such as `list`, will usually not call `Model#commitAddressBook()`, `Model#undoAddressBook()` or `Model#redoAddressBook()`. Thus, the `addressBookStateList` remains unchanged.
+_Figure 2. Class Diagram of the Find Feature_
```

Comment:
Perhaps you should split this into 2 separate diagrams? It may be slightly difficult to see where the arrows are pointing to.

Reply:
I fixed the diagrams in [this](https://5dafe94972d6d40007016365--algobase.netlify.com/developerguide#find-problem-feature) preview. Can you take a look?

Reply:
it's very nice that you can have conversations with teammates hhh